### PR TITLE
Merge healpix mapmaker

### PR DESCRIPTION
**Pull request by earosenberg**
Add healpix support to the sotodlib mapmaking module.
Includes some re-factoring of demod_mapmaker to minimize code duplication

### CODE REVIEW

**Comment by chervias** [comment]
We changed the `tod_grouping` file to `obs_grouping`. Can you fix that please?

File: sotodlib/coords/pmat.py
Comment (mhasself) [inline sotodlib/coords/pmat.py:446]: This comment is associated with the defn of interpol_kw so move it too.

File: sotodlib/coords/pmat.py
Comment (mhasself) [inline sotodlib/coords/pmat.py:483]: This line is healpix specific, doesn't it break non-healpix projector?

File: sotodlib/coords/pmat.py
Comment (mhasself) [inline sotodlib/coords/pmat.py:138]: You can just reuse self.tiled, right?

It is a bug (which I have fixed in a brack somewhere) that this constructor doesn't set `self.tiled = None`.  So I suggest you do that, and than update it to self.hp_geom.nside_tile if conditions support that.

There are a lot of places where you need to do things differently depending on rectpix vs. healpix.  So consider inventing enumerator `self.pix_scheme`, with values none, rectpix, healpix, and choose branches based on that rather than depending on what things are None.

File: sotodlib/coords/pmat.py
Comment (mhasself) [inline sotodlib/coords/pmat.py:145]: Update docstring.

**Review by mhasself — CHANGES_REQUESTED** [review]
A few comments, but then I realized this is based on the SAT mapmaking branch, not master.

Do we want this all to go into master now, or do you want to do healpix -> SAT_mapmaking first, or healpix -> master first?

**Comment by earosenberg** [comment]
> A few comments, but then I realized this is based on the SAT mapmaking branch, not master.
> 
> Do we want this all to go into master now, or do you want to do healpix -> SAT_mapmaking first, or healpix -> master first?

My thought was that this is similar in spirit to #891, so could go directly to master (and perhaps be merged from there into SAT-mapmaking). But I suppose if that is still the main branch for mapmaking the most important thing is to put it there, so we could go the other way. Happy to go with whatever you and @chervias think is easiest/best.

File: sotodlib/coords/pmat.py
Comment (earosenberg) [reply sotodlib/coords/pmat.py:446]: Done

File: sotodlib/coords/pmat.py
Comment (earosenberg) [reply sotodlib/coords/pmat.py:483]: It did; I had neglected to test tiled rectpix projectors. Fixed now

File: sotodlib/coords/pmat.py
Comment (earosenberg) [reply sotodlib/coords/pmat.py:138]: Thanks for the suggestion, I have implemented it.
`self.tiled` is implemented here as a property toward the bottom of the file, so it can't be directly initialized to None. That should now work nicely with healpix as well though, so I have merged them (and geom / hp_geom)

**Comment by mhasself** [comment]
@earosenberg  I think it will be fine to PR this towards master rather than a mapmaking branch.

Can you rebase + squash the commit history here, please? Currently it's deeply intertwined with other branches, but then includes reversion of some changes from those branches; so it would be better to present a simpler view of what's proposed here.

**Comment by chervias** [comment]
I deleted the PR for the SAT_mapmaking branch, we are doing a refactor and we'll go through another route, so please merge into master.

File: sotodlib/coords/pmat.py
Comment (earosenberg) [reply sotodlib/coords/pmat.py]: Done

File: sotodlib/coords/pmat.py
Comment (mhasself) [inline sotodlib/coords/pmat.py:345]: Two lines, please.  Or you can make this a one-liner with an idempotent form of the function, `weights_map = hp_utils.as_compressed(weights_map)`.

File: sotodlib/coords/pmat.py
Comment (mhasself) [inline sotodlib/coords/pmat.py:548]: Make this 4 lines.  If you hadn't broken the alignment with the `else:` line I might have let this one slide. :)

The one liner is acceptable... `return (map.tiles if ... else map)`

File: sotodlib/coords/pmat.py
Comment (mhasself) [inline sotodlib/coords/pmat.py:623]: You might not need this function any more, or it might become private.  The docstring you have needs punctuation and refilling.  Or, remove it entirely; when a docstring for a short private function just restates in words what the code clearly manifests ... that's not super helpful and it can be left out.

File: sotodlib/mapmaking/demod_mapmaker.py
Comment (mhasself) [inline sotodlib/mapmaking/demod_mapmaker.py:293]: I don't think it will be sustainable to have a separate class just to support Healpix projection.  Or more to the point: this class and DemodSignalMap contain too much complex, duplicated code.  We have to find a re-factoring that reduces such duplication.

File: sotodlib/mapmaking/demod_mapmaker.py
Comment (mhasself) [inline sotodlib/mapmaking/demod_mapmaker.py:70]: `det_weights argument` is not documented anywhere, but has rich semantics...

**Review by mhasself — CHANGES_REQUESTED** [review]
Thanks for pushing this forward!  Lots of comments.  We might need to brainstorm the right way to enrich demod_mapmaker.py classes to support healpix without too much copy paste.

We don't have enough unit testing for Pmat... that's beyond your scope, I guess -- but if you want to take a stab at inventing that, it would be welcome.

File: sotodlib/mapmaking/demod_mapmaker.py
Comment (iparask) [reply sotodlib/mapmaking/demod_mapmaker.py:293]: @mhasself these are very similar to the Signal class that are in the mapmaker.py. 

That can be the initial abstraction and move from there

Comment (mhasself) [reply sotodlib/mapmaking/demod_mapmaker.py:293]: Ya, that's certainly the origin of DemodSignalmap... but stripped down for simple filter+bin use.  We should indeed be unforking these classes, not forking them more.

Comment (chervias) [reply sotodlib/mapmaking/demod_mapmaker.py:293]: I agree with Matthew, I don't think we should have a new class.

Comment (earosenberg) [reply sotodlib/mapmaking/demod_mapmaker.py:293]: Happy to refactor this. Perhaps a single class with two separate constructors `DemodSignalMap.for_rectpix` and `DemodSignalMap.for_healpix`, following the model of `pmat.P`? The base `__init__` would stay the same but with all args turned into kwargs and expanded with healpix kwargs, so existing code calling it will not break. 
Then I should be able to set a `pix_scheme` flag and navigate the branches straightforwardly as in `pmat`. 

If this sounds acceptable I'll proceed with the implementation, let me know of any other suggestions.

Comment (mhasself) [reply sotodlib/mapmaking/demod_mapmaker.py:293]: That sounds good to me!

**Comment by earosenberg** [comment]
OK addressed above comments, did the re-factor, and added some documentation

File: sotodlib/mapmaking/demod_mapmaker.py
Comment (mhasself) [inline sotodlib/mapmaking/demod_mapmaker.py:241]: Pass parameters by name, please!

File: sotodlib/mapmaking/demod_mapmaker.py
Comment (mhasself) [inline sotodlib/mapmaking/demod_mapmaker.py:277]: Pass params by name.

File: sotodlib/mapmaking/demod_mapmaker.py
Comment (mhasself) [inline sotodlib/mapmaking/demod_mapmaker.py:378]: I think we'd be wise to write both the array and some supporting info the projection.  NEST is essential for understanding the data.  NSIDE is helpful.  Eventually we might throw in the celestial frame (galactic / equatorial).

File: sotodlib/mapmaking/demod_mapmaker.py
Comment (mhasself) [inline sotodlib/mapmaking/demod_mapmaker.py:383]: Same here -- add attributes to this dataset with important projection info.

**Review by mhasself — CHANGES_REQUESTED** [review]
Ok!  I think this is looking good.  A few comments about docs mostly.

Thanks for extending the existing tests -- that paves the way for more / better in the future.

Can you remind me what the use of `qp_kwargs` has been?  It is a bit suspicious to have that, and `det_weights`, added to the `DemodMapmaker.add_obs`.  You'd think that noise_model could somehow capture the effects of det_weights, and geom (or the class constructor, more generally) could handle whatever qp_kwargs is doing.  I have stopped looking into this in the spirit of getting this review in in finite time, but perhaps you and @chervias could consider that, from the perspective of eventually unforking the mapmaker & signal classes.

**Comment by earosenberg** [comment]
> Can you remind me what the use of `qp_kwargs` has been? 

I am happy to leave or remove this as you and @chervias think. This was put in place to enable changing the settings of qpoint -- primarily so they could match TOAST settings. The point was to allow to_map match an input map that was scanned with TOAST at default settings. I think this is potentially useful, but maybe a sufficiently niche application that it's not worth complicating the code with.

>You'd think that noise_model could somehow capture the effects of det_weights

I think this is correct. The main function of `det_weights` now is to disable inverse variance weighting for testing, but this can easily be achieved with Nmat. I'll remove it, thanks for the suggestion.

File: sotodlib/mapmaking/demod_mapmaker.py
Comment (earosenberg) [reply sotodlib/mapmaking/demod_mapmaker.py:383]: Done, added ordering and nside as attributes

File: sotodlib/mapmaking/demod_mapmaker.py
Comment (mhasself) [reply sotodlib/mapmaking/demod_mapmaker.py:378]: Like this, basically:
```
np.save('test.npy', {'nside': 10, 'data': np.zeros(1000)}, allow_pickle=True)
```

Comment (earosenberg) [reply sotodlib/mapmaking/demod_mapmaker.py]: Got it, thanks for the help!

**Comment by earosenberg** [comment]
That should be the most recent comments done -- thanks for the continued feedback. For `qp_kwargs` I thought it seemed sensible to remove it from `demod_mapmaker` but leave it in `pmat`, but happy to change that if there are any objections.

By the way, I also have a unit test file for `healpix_utils`, mostly checking that the conversions between different healpix formats (tiled, compressed, full) are working right. Wasn't sure if you want this to be included, but will add it if so.

**Comment by chervias** [comment]
Besides fixing the conflict, is there anything stopping us from merging? I need this to be merged first so that I can start moving towards merging #973.

**Comment by earosenberg** [comment]
Not from my side -- was waiting for the go-ahead before doing a final rebase and resolving the conflict.

**Comment by mhasself** [comment]
> Not from my side -- was waiting for the go-ahead before doing a final rebase and resolving the conflict.

This is good to go other than the conflict.  Please rebase, fix conflict, squash ... in some order ... and I'll approve.

**Comment by earosenberg** [comment]
Rebase/conflict/squash all done. 
NB the new `test_pmat.py` is failing post-rebase on tiled rectpix maps. It seems that the new `enmapify` in `remove_weights` and `invert_weights` doesn't work for the tiled case. I can remove those lines from the test if it needs to pass before merging, but it seemed like an issue to address outside this PR.

**Comment by earosenberg** [comment]
> Ok ... ya, just suppress that test (comment it out) and we can fix it separately.

Great, done

**Review by mhasself — APPROVED** [review]
Quick merge it before something else breaks

**Pull request merged**
